### PR TITLE
2311: pt4 -Added prefixed `descendant_` and `ancestor_` url parameters for /relations endpoint

### PR DESCRIPTION
**Pull request by rohan-bes**
### Changes:

Allowed for specifying of fields and filtering of ancestors/descendants in `/relations` endpoint, eg:
```
http://{{host}}/{{version}}/hierarchy/laos_schools/LA/relations?ancestor_type=district&descendant_type=school&field=code&descendant_attributes_type=Secondary&groupBy=descendant
=>
All school->district mappings where descendant.attributes.type = 'Secondary'
```

### CODE REVIEW

File: packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+import { formatEntitiesForResponse } from '../format';
+import { EntityResponseObject, FlattenedEntity } from '../types';
+import { RelationsRequest } from './types';
+
+type Pair = {
+  descendant: string;
+  ancestor: string;
+};
+
+type UnArray<T> = T extends (infer U)[] ? U : never;
```
Comment (rohan-bes) [inline packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts:18]: Really wish this was a standard type 😡

File: packages/entity-server/src/routes/hierarchy/relations/middleware.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import { extractFilterFromQuery } from '../middleware/filter';
+import { RelationsRequest, RelationsQuery, Prefix } from './types';
+
+type Strip<T extends string, U extends string> = T extends `${U}_${infer P}` ? P : T;
+const strip = <T extends string, U extends string>(baseString: T, toStrip: U) => {
+  const stripped = baseString.startsWith(`${toStrip}_`)
+    ? baseString.split(`${toStrip}_`)[1]
+    : baseString;
+  return stripped as Strip<T, U>;
+};
```
Comment (rohan-bes) [inline packages/entity-server/src/routes/hierarchy/relations/middleware.ts:18]: I ❤️ TS

Comment (kael89) [reply packages/entity-server/src/routes/hierarchy/relations/middleware.ts:17]: Haha that would make a great T-Shirt!

File: packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            ...(await formatEntitiesForResponse(models, ctx, descendants, descendantFlat)),
+          ];
+    const formattedEntitiesByCode: Record<string, string> = {};
+    [...ancestors, ...descendants].forEach((entity, index) => {
+      formattedEntitiesByCode[entity.code] = formattedEntities[index];
+    });
+    return formattedEntitiesByCode;
+  }
+
+  shouldPerformFastResponse() {
```
Comment (kael89) [inline packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts:68]: Very clean implementation!

File: packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+  async getAncestors(ancestorsWithDescendantsCodes: string[]) {
+    const { entity, hierarchyId, allowedCountries } = this.req.ctx;
+    const { type: ancestorType, filter } = this.req.ctx.ancestor;
+    return entity.type === ancestorType
+      ? [entity]
+      : entity.getDescendants(hierarchyId, {
+          code: ancestorsWithDescendantsCodes,
+          ...filter,
+          country_code: allowedCountries,
```
Comment (kael89) [inline packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts:45]: Question: if the client provides a `country_code` filter, we should use the intersection of `filter.country_code` and `allowedCountries` as the `country_code` condition. I am not sure if this logic is already used to evaluate `allowedCountries`?

File: packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    const formattedEntities =
+      ancestorFlat === descendantFlat
+        ? await formatEntitiesForResponse(models, ctx, [...ancestors, ...descendants], ancestorFlat)
+        : [
+            ...(await formatEntitiesForResponse(models, ctx, ancestors, ancestorFlat)),
+            ...(await formatEntitiesForResponse(models, ctx, descendants, descendantFlat)),
+          ];
+    const formattedEntitiesByCode: Record<string, string> = {};
+    [...ancestors, ...descendants].forEach((entity, index) => {
+      formattedEntitiesByCode[entity.code] = formattedEntities[index];
```
Comment (kael89) [inline packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts:63]: I guess that `formatEntitiesForResponse` guarantees a 1-1 and ordered mapping between the entity input and output?

**Review by kael89 — APPROVED** [review]
I love this PR, the code is easy to follow and extremely clean!

File: packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts (hunk 4 above)
Comment (rohan-bes) [reply packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts:45]: That's a good point, we probably ought to 👍

File: packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts (hunk 5 above)
Comment (rohan-bes) [reply packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts:63]: Correct

File: packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts (hunk 4 above)
Comment (kael89) [reply packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts:45]: It's a bit of a detail and not a big blocker, so happy for it to be addressed in the end, after the refactoring & merging dust has settled

File: packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts [hunk 6]
```diff
@@ -75,51 +75,13 @@ export class EntityRelationsRoute extends Route<RelationsRequest> {
...
-      formattedEntitiesByCode[descendantToAncestor[descendant.code]];
-    const getFormattedDescendant = (descendant: EntityType) => [
-      formattedEntitiesByCode[descendant.code],
-    ];
-    return reduceToDictionary(
-      descendants.filter(getFormattedAncestor),
-      getFormattedDescendant,
-      getFormattedAncestor,
-    );
+  buildMap(pairs: Pair[]) {
```
Comment (kael89) [inline packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts:92]: Amazing simplification in this commit!!

File: packages/entity-server/src/routes/hierarchy/middleware/filter.ts [hunk 7]
```diff
@@ -3,9 +3,23 @@
...
-import { EntityFilter } from '../../../models';
-import { EntityFilterQuery, NestedFilterQueryFields } from '../types';
+import { Writable, ObjectLikeKeys, Flatten } from '../../../types';
+import { EntityFilter, EntityFields } from '../../../models';
+
+const CLAUSE_DELIMITER = ';';
+const FIELD_VALUE_DELIMITER = ':';
+const NESTED_FIELD_DELIMITER = '_';
+const JSONB_FIELD_DELIMITER = '->>';
+const MULTIPLE_VALUES_DELIMITER = ',';
```
Comment (kael89) [inline packages/entity-server/src/routes/hierarchy/middleware/filter.ts:13]: Great abstractions!

File: packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts [hunk 8]
```diff
@@ -35,14 +35,23 @@ export class EntityRelationsRoute extends Route<RelationsRequest> {
...
+        : ancestorsWithDescendantsCodes.filter(code => filterCode === code);
+    }
+
     return entity.type === ancestorType
       ? [entity]
       : entity.getDescendants(hierarchyId, {
-          code: ancestorsWithDescendantsCodes,
           ...filter,
-          country_code: allowedCountries,
+          code: codesToUse,
```
Comment (kael89) [inline packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts]: Good proactive fix!

**Pull request merged**